src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1580 @@
+	if (r < 0)
+		goto fail;
+
+	left = size = lebytes2ushort(count);
+sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"got %d bytes out_ptr=%lx count&=%lx count[0]=0x%02x count[1]=0x%02x, len=0x%04x (%d)", len, (unsigned long) out_ptr, (unsigned long)&count, count[0], count[1], size, size); 
+	out = out_ptr = malloc(size);
+	if (out == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto fail;
+	}
+	for (offset += 2; left > 0; offset += len, left -= len, out_ptr += len) {
+		len = MIN(left, CAC_MAX_CHUNK_SIZE);
+		params[1] = len;
+		r = cac_apdu_io(card, CAC_INS_READ_FILE, HIGH_BYTE_OF_SHORT(offset), LOW_BYTE_OF_SHORT(offset), 
+						&params[0], sizeof(params), &out_ptr, &len);
+		if (r < 0) {
+			goto fail;
+		}
+	}
+	*out_len = size;
+	*out_buf = out;
+	return SC_SUCCESS;
+fail:
+	if (out) 
+		free(out);
+	*out_len = 0;
+	return r;
+}
+
+/*
+ * OLD cac read certificate, only use with CAC-1 card.
+ */
+static int cac_cac1_get_certificate(sc_card_t *card, u8 **out_buf, size_t *out_len)
+{
+	u8 buf[CAC_MAX_SIZE];
+	u8 *out_ptr;
+	size_t size = 0;
+	size_t left = 0;
+	size_t len, next_len;
+	sc_apdu_t apdu;
+	int r;
+
+
+	/* get the size */
+	size = left = *out_buf ? *out_len : sizeof(buf);
+	out_ptr = *out_buf ? *out_buf : buf;
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, CAC_INS_GET_CERTIFICATE, 0, 0 );
+	next_len = MIN(left, 100);
+	for (; left > 0; left -= len, out_ptr += len) {
+		len = next_len;
+		apdu.resp = out_ptr;
+		apdu.le = len;
+		apdu.resplen = left;
+
+		r = sc_transmit_apdu(card, &apdu);
+		if (r < 0) {
+		 	break;
+		}
+		/* in the old CAC-1, 0x63 means 'more data' in addition to 'pin failed' */
+		if (apdu.sw1 != 0x63)  {
+			/* we've either finished reading, or hit an error, break */
+			r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+			left -= len;
+			break;
+		}
+		next_len = MIN(left,apdu.sw2);
+	}
+	if (r < 0) {
+		return r;
+	}
+	r = size - left;
+	if (*out_buf == NULL) {
+		*out_buf = malloc(r);
+		if (*out_buf == NULL) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+		memcpy(*out_buf, buf, r);
+	}
+	*out_len = r;
+	return r;
+}
+
+/* create a fake tag/length file for cac1 cards based on the val_len */
+static int cac_cac1_get_cert_tag(sc_card_t *card, size_t val_len, u8 **tlp, size_t *tl_len_p) 
+{
+	static const u8 cac_cac1_cert_tag[] = { CAC_TAG_CERTINFO, 1, CAC_TAG_CERTIFICATE, 0xff, 0, 0 };
+	u8 *tl;
+	size_t tl_len;
+	tl_len = sizeof(cac_cac1_cert_tag);
+	tl = malloc(tl_len);
+	if (tl == NULL) {
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	memcpy(tl, cac_cac1_cert_tag, tl_len);
+	val_len--; /* one byte is CERTINFO */
+	if (val_len < 0xff) {
+		tl_len -= 2;
+		tl[tl_len-1] = (u8)val_len;
+	} else {
+		tl[tl_len-2] = (u8)val_len & 0xff;
+		tl[tl_len-1] = (u8)(val_len >> 8) & 0xff;
+	}
+	*tlp = tl;
+	*tl_len_p = tl_len;
+	return SC_SUCCESS;
+}
+		
+/* get the tlv length return that len and return the length of the tl header. Don't overrun the buffer */
+static size_t cac_get_len(u8 *tl_ptr, size_t tl_len, size_t *tl_head_len) 
```

Comment:
Actually the Simple TLV is used also here, which is a reason why the `asn1.h` is not useful (the length parsing is a bit different).